src/optimize.rs

```diff
@@ -0,0 +1,1049 @@
+                            new_insts.push(CowInst::Owned(wasm_encoder::Instruction::I64Add));
+                            new_insts.push(CowInst::Owned(wasm_encoder::Instruction::GlobalSet(
+                                entry.counters.unwrap().1,
+                            )));
+                        }
+
+                        new_insts.push(CowInst::Owned(wasm_encoder::Instruction::LocalGet(
+                            temp_callee_local,
+                        )));
+                        new_insts.push(CowInst::Owned(wasm_encoder::Instruction::CallIndirect {
+                            ty: info.type_index,
+                            table: info.table_index,
+                        }));
+                        new_insts.push(end);
+                    }
+
+                    on_stack.remove(&entry.defined_func_index);
+                    stack.pop();
+                    continue;
+                }
+            };
+
+            match op {
+                Operator::CallIndirect {
+                    type_index,
+                    table_index,
+                    table_byte: _,
+                } => {
+                    entry.call_site_index += 1;
+                    if let Some(new_entry) = self.try_enqueue_for_winlining(
+                        context,
+                        &on_stack,
+                        &mut locals,
+                        &mut num_locals,
+                        &mut new_insts,
+                        temp_callee_local,
+                        entry.call_site_index - 1,
+                        table_index,
+                        type_index,
+                    )? {
+                        on_stack.insert(new_entry.defined_func_index);
+                        stack.push(new_entry);
+                    } else {
+                        new_insts.push(CowInst::Owned(wasm_encoder::Instruction::CallIndirect {
+                            ty: type_index,
+                            table: table_index,
+                        }));
+                    }
+                }
+
+                // `local.{get,set,tee}` instruction's need their local index adjusted.
+                Operator::LocalGet { local_index } if entry.locals_delta != 0 => {
+                    new_insts.push(CowInst::Owned(wasm_encoder::Instruction::LocalGet(
+                        local_index + entry.locals_delta,
+                    )))
+                }
+                Operator::LocalSet { local_index } if entry.locals_delta != 0 => {
+                    new_insts.push(CowInst::Owned(wasm_encoder::Instruction::LocalSet(
+                        local_index + entry.locals_delta,
+                    )))
+                }
+                Operator::LocalTee { local_index } if entry.locals_delta != 0 => {
+                    new_insts.push(CowInst::Owned(wasm_encoder::Instruction::LocalTee(
+                        local_index + entry.locals_delta,
+                    )))
+                }
+
```

Comment:
As noted earlier -- we'll need to handle `Return`s by creating a branch to the outermost level of the current inlined function body, I think. We'll need to keep a block-depth counter tracked through block/loop/if/else/end ops to know how far out to jump.